### ISSUE TITLE
darcs completion does not recognize multi-level commands

### ISSUE THREAD

**Issue by bfrk** [fish-shell/fish-shell#3979]
For quite some time now, darcs has subcommands that have themselves subcommands. For instance, here is the output of `darcs show --help`:

```
Usage: darcs show [OPTION]...
Show information about the given repository.

Subcommands:
  contents       Outputs a specific version of a file.
  dependencies   Generate the graph of dependencies.
  files          Show version-controlled files in the working tree.
  index          Dump contents of working tree index.
  pristine       Dump contents of pristine cache.
  repo           Show repository summary information
  authors        List authors by patch count.
  tags           Show all tags in the repository.
  patch-index    Check integrity of patch index

Options:
   --help          show a brief description of the command and its options
   --list-options  show plain list of available options and commands, for auto-completion
   --disable       disable this command

Use the `--help` option with the subcommands to obtain help for
subcommands (for example, `darcs show files --help`).
```

It would be nice if the generic vcs completion script builder could be extended to cover this possibility.

My fish version is 2.2.0 but the problem is also present with freshly re-created darcs completions, using ./build_tools/make_darcs_completions.fish of the current HEAD.

**Comment by faho** [comment]
You know, we've kinda moved away from the generation script, since it doesn't actually handle a bunch of cases.

I don't use darcs and getting to grips with these kinds of tools takes a while, so I'd appreciate it if a darcs user were to step up and fix it directly in share/completions/darcs.fish.

**Comment by bfrk** [comment]
> You know, we've kinda moved away from the generation script, since it doesn't actually handle a bunch of cases.

Ah well. Handling all cases for something as complex as a mature VCS is difficult. But if the alternative is to maintain every option of every command manually, I would rather accept something incomplete. For illustration, the zsh completion for git is a monster of 7000 code lines. A huge user community such as git's can perhaps afford to write and maintain such a thing, but for much smaller ones such as darcs this is out of the question. Which is is why I found the approach with the generic builder script so interesting.

Our approach to getting reasonably well working shell completion without too much maintenance effort so far has been to offer support from darcs itself, using the special option --list-options (which actually lists all possible completions, including non-option arguments). This approach has its limits, for instance we cannot (currently) offer completions for option arguments. On the other hand, the zsh completion script in its latest version has just 40 lines including comments and works quite well in spite of the above mentioned limitations (which is to some extent due to the fact that most darcs commands are interactive by default and thus options are used less frequently).

Does fish offer something akin to zsh's _multi_parts function? This is very useful to complete files in "virtual" file trees that may not actually exist on the disk.

**Comment by faho** [comment]
>Does fish offer something akin to zsh's _multi_parts function? This is very useful to complete files in "virtual" file trees that may not actually exist on the disk.

I have no idea. What does that function do? Or how do you expect to use it?

**Comment by bfrk** [comment]
It has many features (too many, like everything in zsh) but the important one is this: let's say I have an array variable named `paths` with a large number of all possible file path completions. Like what you get with `find <some dir>` only that the files may not actually exist (e.g. the user has removed a file but not yet recorded the change). Then the command

`      _multi_parts / paths`

adds completions in such a way that we are immediately offered only the completions up to (including) the next '/' (or end of string), just as with normal file path completion. This includes adding a space at the end only if the match is complete (i.e. we arrived at a file or an empty directory).

The separator `/` above is a parameter (so you can use this function to complete other kinds of hierarchies, like URLs or `user@host`pairs etc), as is the name of the array. An additional feature is that it also completes incomplete parts separately e.g. `/e/pas` would be completed to `/etc/passwd` (assuming `/etc/passwd` is a member of `paths`), but that is not so important.

**Comment by faho** [comment]
>An additional feature is that it also completes incomplete parts separately e.g. /e/pas would be completed to /etc/passwd (assuming /etc/passwd is a member of paths), but that is not so important.

In fish, that's just part of the completion logic everywhere. The script generates something, and then fish does some fuzzy-matching on it.

>_multi_parts / paths

>adds completions in such a way that we are immediately offered only the completions up to (including) the next '/' (or end of string), just as with normal file path completion. This includes adding a space at the end only if the match is complete (i.e. we arrived at a file or an empty directory).

>The separator / above is a parameter (so you can use this function to complete other kinds of hierarchies, like URLs or user@hostpairs etc), as is the name of the array.

We don't have anything like this _specifically_ but we do have completions that work like that. I'll have to look for examples.

What exactly do you want to use this for?

**Comment by bfrk** [comment]
> What exactly do you want to use this for?

Version control, darcs in particular. Complete a file path that may or may not exist on disk, but does exist in the memory of the VCS. The idea is to query the VCS for all possible paths and use the list as input to the completion. Example: show me all patches that touch a certain file, inside a bare repo i.e. w/o working tree.

**Comment by krader1961** [comment]
Removing the "needs-more-info" label since as far as I can tell this simply requires someone to improve the *share/completions/darc.fish* script.

**Comment by faho** [comment]
Fixed by d4f5689eaa2f7c7f517649d43ef45d01a4f8248c in 3.0, I believe.
